### PR TITLE
[DEVEX-109] Add Release workflow for Infrastructure

### PR DESCRIPTION
**Pull request by Krusty93**
Set a workflow that:
- is triggered when code on `main` branch is pushed under `infra/resources` path
- consists in two jobs:
  - the first one:
    - performs a `terraform plan` and upload results as artifacts that will stay available for 7 days
    - plan output is also available in logs
    - runs with the CI identity
  - the second job:
    - downloads the artifact and apply the changes
    - runs with the CD identity
    - is meant to be explicitly approved by a repository CODEOWNER, after reviewing plan output

Considerations:
- Running `terraform apply` using the output of the command `terrafrom plan` had run in the previous job guarantees that the effects are exactly the same as shown before. If the underlying status changes for any reason, the apply fails and the workflow must be restarted and a new plan must be generated (and read)
- Concurrency checks have been introduced, in order to create a queue of workflows and avoid overlapping deployments. This also forces to cancel any in-progress workflow stuck before running it again

Design and organization are a consequence of decisions made on #11

### CODE REVIEW

**Review by lucacavallaro — APPROVED** [review]

File: .github/workflows/infra_apply.yaml [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    steps:
+
+      # Set the directory where the Terraform files are located
+      # The directory the value is then available in ${{ steps.directory.outputs.dir }}
+      - name: Set directory
+        id: directory
+        run: |
+          set -e
+          if [ -z "${{ inputs.environment }}" ] || [ -z "${{ inputs.region }}" ]; then
```
Comment (gunzip) [inline .github/workflows/infra_apply.yaml:60]: do these need to be sanitized?

File: .github/workflows/infra_apply.yaml [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            echo "dir=${{ inputs.base_path}}/${{ inputs.environment }}/${{ inputs.region }}" >> $GITHUB_OUTPUT
+          fi
+
+      - uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        name: Checkout
+
+      - name: Set Environment Variables
+        if: ${{ inputs.env_vars }}
+        run: |
+          for i in "${{ inputs.env_vars }}"
```
Comment (gunzip) [inline .github/workflows/infra_apply.yaml:76]: ditto

File: .github/workflows/infra_apply.yaml [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+      - name: Set directory
+        id: directory
+        run: |
+          set -e
+          if [ -z "${{ inputs.environment }}" ] || [ -z "${{ inputs.region }}" ]; then
+            echo "Both environment and region must be provided."
+            exit 1
+          else
+            # The directory is expected to be in the format
+            #  infra/resources/${{ inputs.environment }}/${{ inputs.region }}
```
Comment (lucacavallaro) [inline .github/workflows/infra_apply.yaml:65]:
```suggestion
            #  ${inputs.base_path}/${{ inputs.environment }}/${{ inputs.region }}
```

(?)

Comment (lucacavallaro) [reply .github/workflows/infra_apply.yaml:65]: imho this comment is not necessary, the code is already clear enough. I would just leave the example (infra/westeurope ...)

**Review by gunzip — APPROVED** [review]

**Pull request merged**